File: server/app/routers/graph.py
"""
server/app/routers/graph.py
W-P-004 关系图谱（只读语义层；未 BUILD 返回空图 available:false，不 500）。

P7：同一取数端点参数化——
  edge_kinds：逗号分隔 link 名白名单（资金链路图=transfers；
              关系图谱缺省全部），不新增取数 API；
  clue_id：仅高亮该线索；缺省时高亮最新产物中本角色可见的全部线索。
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query

from core.access import can_see_jian_types

from server.app import graph_view
from server.app.clues_view import _load_raw
from server.app.deps import WebContext, get_ctx, get_principal
from server.app.envelope import ok
from server.app.ontology_meta import jian_clearances
from server.app.routers.cases import _get_owned_case
from server.app.security import Principal

router = APIRouter(tags=["graph"])


def _parse_edge_kinds(raw: str | None) -> list[str]:
    if not raw:
        return []
    return [x.strip() for x in raw.split(",") if x.strip()]


@router.get("/cases/{case_id}/graph")
def get_graph(case_id: str,
              node_limit: int = Query(300, ge=1, le=1000),
              edge_limit: int = Query(500, ge=1, le=3000),
              edge_kinds: str | None = Query(
                  None, description="逗号分隔 link 名；缺省全部入图边"),
              clue_id: str | None = Query(
                  None, description="仅高亮该线索；缺省高亮全部可见线索"),
              p: Principal = Depends(get_principal),
              ctx: WebContext = Depends(get_ctx)):
    """实体关系图：节点=entity 型对象按度数采样，边=双侧带 ref 链接。"""
    _get_owned_case(case_id, p, ctx.cases)
    case = ctx.repo.get_case(case_id)
    data = {"available": False,
            "truncated": {"nodes": False, "edges": False, "dropped_edges": 0},
            "edge_kinds": [],
            "highlights": {"clue_count": 0, "nodes": 0, "edges": 0},
            "nodes": [], "edges": []}
    kinds = _parse_edge_kinds(edge_kinds)
    store = None
    state = None
    try:
        try:
            store = ctx.factory.for_case(case_id, mode="read")
            # P8 人验图像证据（state.sqlite）——版本库不含 runtime 表，
            # 读 state 注入 assemble_graph（核验后自动入图）。
            from server.app.store.state_store import StateStore
            state = StateStore(case_id,
                               ctx.factory.case_dir(case_id) / "state.sqlite")
            image_rows = state.list_image_evidence(case_id=case_id)
            # 高亮线索来自案件线索产物（随版本不可变），按角色间类可见性过滤，
            # 不可见线索不进高亮（防泄漏）。
            snapshot_root = ctx.cases.snapshot_ontology_root(case_id)
            clearances = jian_clearances(case.pack_id, snapshot_root)
            raws, _ = _load_raw(ctx.factory.case_dir(case_id), None)
            highlight_clues: list[dict] = []
            for r in raws:
                if clue_id is not None and r.get("clue_id") != clue_id:
                    continue
                jts = r.get("jian_types") or []
                if not can_see_jian_types(jts, role=p.role,
                                          jian_clearances=clearances):
                    continue
                highlight_clues.append(r)
            data = graph_view.assemble_graph(
                conn=store.read_conn, pack=case.pack_id,
                base_dir=snapshot_root,
                node_limit=node_limit, edge_limit=edge_limit,
                edge_kinds=kinds or None,
                highlight_clues=highlight_clues,
                image_evidence_rows=image_rows)
        except FileNotFoundError:
            pass  # 未 BUILD（无版本文件）→ 空图
        except ValueError as exc:
            # 未声明 edge_kinds 等参数错误 = 请求方问题（400，不 500）
            raise HTTPException(status_code=400, detail=str(exc))
    finally:
        if store is not None:
            store.close()
        if state is not None:
            state.close()
    return ok(data, data_version=ctx.repo.current_version(case_id))

File: frontend/src/api/endpoints/graph.ts
import { api } from '../client'
import { noteDataVersion } from '../query-keys'

// 知识图谱（server/app/routers/graph.py + graph_view.py 契约）。
// GET /cases/{cid}/graph?node_limit=&edge_limit=&edge_kinds=&clue_id=
//   节点按度数 top-N 采样（非全量）；未 BUILD/无实体表 → available:false 空结构（不 500）。
//   edge_kinds：同一取数的 link 白名单投影（资金链路图=transfers）。
//   被线索 evidence_refs 引用的节点/边带 hit/clue_ids 高亮。

export interface GraphNode {
  /** id 域 "<object>:<pk>"，如 person:123 */
  id: string
  label: string
  type: string
  type_title: string
  /** 命中间名（因/内/反/死/生），用于节点着色 */
  jian: string[]
  /** 被线索 evidence_refs 直接引用 */
  hit: boolean
  /** 命中该节点的线索 ID（可溯源） */
  clue_ids: string[]
}

export interface GraphEdgeProps {
  amount?: string
  date?: string
}

export interface GraphEdge {
  source: string
  target: string
  label: string
  type: string
  /** 被线索 evidence_refs 引用 */
  hit: boolean
  clue_ids: string[]
  /** build_sql 实际输出中的附加属性（amount/date 白名单） */
  props: GraphEdgeProps
}

export interface EdgeKind {
  name: string
  title: string
  /** 案件语义层是否存在该边表 */
  present: boolean
}

export interface GraphHighlights {
  clue_count: number
  nodes: number
  edges: number
}

export interface GraphDto {
  available: boolean
  truncated: { nodes: boolean; edges: boolean; dropped_edges: number }
  edge_kinds: EdgeKind[]
  highlights: GraphHighlights
  nodes: GraphNode[]
  edges: GraphEdge[]
}

export interface GraphQuery {
  nodeLimit?: number // 1–1000，默认 300
  edgeLimit?: number // 1–3000，默认 500
  edgeKinds?: string[] // link 名白名单；缺省全部
  clueId?: string // 仅高亮该线索；缺省全部可见线索
}

export const graphApi = {
  async get(caseId: string, query: GraphQuery = {}): Promise<GraphDto> {
    const params = new URLSearchParams()
    if (query.nodeLimit) params.set('node_limit', String(query.nodeLimit))
    if (query.edgeLimit) params.set('edge_limit', String(query.edgeLimit))
    if (query.edgeKinds?.length) params.set('edge_kinds', query.edgeKinds.join(','))
    if (query.clueId) params.set('clue_id', query.clueId)
    const qs = params.toString()
    const res = await api.get<GraphDto>(
      `/cases/${encodeURIComponent(caseId)}/graph${qs ? `?${qs}` : ''}`,
    )
    noteDataVersion(caseId, res.dataVersion)
    return res.data
  },
}
